### DISCUSSION TITLE
Modify confidence level for confusion matrix

### DISCUSSION THREAD

**Discussion by ayeshans** [ultralytics/ultralytics#6730] [Q&A]
👋 Hi there! I am using ultralytics yolov8 for an object detection task. I am noticing some discrepancies in the DetMetrics object called metrics, generated using this code:

```python
from ultralytics import YOLO
...
local_path_model_weights = ...
path_data_yaml = ... 
model = YOLO(local_path_model_weights)
metrics = model.val(data=path_data_yaml, conf=0.15, split='test')
```

I am accessing the class-wise precision and recall using `[metrics.box](http://metrics.box).p` and `metrics.box.r` and using these to compute weighted precision, recall and f1-score. However, when I compute class-wise precision and recall using the `metrics.confusion_matrix`, the values of precision and recall are different from the `metrics.box` ones. I realized that this is since `metrics.confusion_matrix.conf` is .25, while presumably the box values use conf .15 (since this is what is passed to model.val()). Since I passed .15, shouldn’t the confusion matrix also use this conf value? How can I make the conf value consistent between `p` and `confusion_matrix`?

**Comment by pderrenger** [comment]
@ayeshans hello! 👋

The discrepancy you're observing between the precision and recall values from `metrics.box` and `metrics.confusion_matrix` is likely due to the different confidence thresholds being applied. When you pass the `conf` argument to `model.val()`, it should indeed set the confidence threshold consistently across all metrics calculated during validation, including the confusion matrix.

To ensure that the confusion matrix uses the same confidence threshold as the one you've specified, you can directly set the `conf` parameter in the `model.val()` method as you've done. If the confusion matrix is not reflecting the specified confidence threshold, this could be a bug or an oversight in the implementation.

Here's what you can do to troubleshoot and resolve this issue:

1. Double-check that you are using the latest version of the Ultralytics YOLOv8 repository, as updates may have addressed this issue.

2. Ensure that there are no overrides or hard-coded values for the confidence threshold within the `model.val()` method or the confusion matrix calculation that could be causing this discrepancy.

3. If the issue persists, consider raising an issue on the Ultralytics YOLOv8 repository with a detailed description of the problem, including the code snippet you've provided and any additional context that could help in diagnosing the issue.

In the meantime, you can manually calculate the confusion matrix using the predictions filtered by your desired confidence threshold to ensure consistency with the other metrics.

I hope this helps! If you have further questions or need assistance, feel free to reach out. 😊
